src/tainting/Dataflow_tainting.ml

```diff
@@ -421,45 +421,65 @@ and eval_label_args ~labels args =
-let findings_of_tainted_sink env taints (sink : T.sink) : T.finding list =
+let taints_satisfy_requires taints expr =
   let labels = labels_of_taints taints in
+  eval_label_requires ~labels expr
+
+(* Potentially produces a finding from incoming taints to a sink.
+   Note that, while this sink has a `requires` and incoming labels,
+   we decline to solve this now!
+   We will figure out how many actual Semgrep findings are generated
+   when this information is used, later.
+*)
+let findings_of_tainted_sink env taints (sink : T.sink) : T.finding option =
+  (* We cannot check whether we satisfy the `requires` here.
+     This is because this sink may be inside of a function, meaning that
+     argument taint can reach it, which can only be instantiated at the
+     point where we call the function.
+     So we record the `requires` within the taint finding, and evaluate
+     the formula later, when we extract the PMs
+  *)
   let sink_pm, ts = T.pm_of_trace sink in
-  let req = eval_label_requires ~labels ts.sink_requires in
-  (* TODO: With taint labels it's less clear what is "the source",
-     * in fact, there could be many sources, each one providing a
-     * different label. Here these would be reported as different
-     * findings... We would need to have multiple sources in `SrcToSink`!
-     * And `ArgtoSink` needs to carry the other taint that reaches the
-     * sink besides the argument. *)
-  taints |> Taints.elements
-  |> List.filter_map (fun (taint : T.taint) ->
-         let tokens = List.rev taint.tokens in
-         match taint.orig with
-         | Arg i ->
-             (* We need to check the label and unifiability requirements
-                at the call site. *)
-             Some (T.ArgToSink (i, tokens, sink))
-         | Src source ->
-             if req then
-               let src_pm, _ = T.pm_of_trace source.call_trace in
-               let* merged_env =
-                 merge_source_sink_mvars env sink_pm.PM.env src_pm.PM.env
-               in
-               Some (T.SrcToSink { source; tokens; sink; merged_env })
-             else None)
+  (* We filter out all taints which do not have unifying mvars, if they
+     exist, and the option is set. This is because we do not want to
+     deny a match if there is a non-unifying taint source, if that source
+     is irrelevant to producing the match.
+  *)
+  let taints_and_bindings_with_unifying_mvars =
+    Taints.elements taints
+    |> List.filter_map (fun t ->
+           let bindings =
+             match t.T.orig with
+             | T.Arg _ -> []
+             | Src source ->
+                 let src_pm, _ = T.pm_of_trace source.call_trace in
+                 src_pm.PM.env
+           in
+           match merge_source_sink_mvars env sink_pm.PM.env bindings with
+           | None -> None
+           | Some _ -> Some (t, bindings))
+  in
+  let* merged_env =
+    taints_and_bindings_with_unifying_mvars |> List.concat_map snd
```

Comment:
What if we have two sources A and B, and both can unify with the sink, but we can't unify A with B?